### PR TITLE
upgrade opc ua nuget package to  version 1.4.366.31-preview

### PR DESCRIPTION
**Pull request by peterbryntesson**
Work in progress - no need to review this yet

### CODE REVIEW

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs [hunk 1]
```diff
@@ -1731,7 +1732,7 @@ private class DummyDialog : IApplicationMessageDlg {
                 DiscoveryUrls = null
             };
             endpoints = _endpoints;
-            return new List<Task> { Task.CompletedTask };
+            return new List<ServiceHost>(); // TODO: We should probably create a service host here - but is this code really used?
```
Comment (simonjaeger) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs:1735]: This is odd. Can we check what to return here_

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj [hunk 2]
```diff
@@ -51,11 +51,11 @@
...
-        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Configuration.Debug" Version="1.4.365-gfc341ee8c5" />
-        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Server.Debug" Version="1.4.365-gfc341ee8c5" />
-        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Client.Debug" Version="1.4.365-gfc341ee8c5" />
-        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Client.ComplexTypes.Debug" Version="1.4.365-gfc341ee8c5" />
+        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Core.Debug" Version="1.4.366.31-preview" />
+        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Configuration.Debug" Version="1.4.366.31-preview" />
+        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Server.Debug" Version="1.4.366.31-preview" />
+        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Client.Debug" Version="1.4.366.31-preview" />
+        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Client.ComplexTypes.Debug" Version="1.4.366.31-preview" />
       </ItemGroup>
```
Comment (simonjaeger) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj:59]: Change for Release mode as well in the .csproj.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Encoders/JsonEncoderEx.cs [hunk 3]
```diff
@@ -1320,6 +1320,56 @@ public enum JsonEncoding {
...
+            try {
+                var arrayLen = matrix.Dimensions[dim];
+                if (dim == matrix.Dimensions.Length - 1) {
+                    // Create a slice of values for the top dimension
+                    var copy = Array.CreateInstance(
+                        matrix.Elements.GetType().GetElementType(), arrayLen);
+                    Array.Copy(matrix.Elements, index, copy, 0, arrayLen);
+                    // Write slice as value rank
+
+#if false // TODO is this really needed when we use the JsonWriter?
```
Comment (simonjaeger) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Encoders/JsonEncoderEx.cs:1351]: Remove it. But we should have tests... any tests from the reference implementation we can add with this PR?

**Review by simonjaeger — CHANGES_REQUESTED** [review]

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj (hunk 2 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj:59]: Fixed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Encoders/JsonEncoderEx.cs (hunk 3 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Encoders/JsonEncoderEx.cs]: I'll take a look

Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Encoders/JsonEncoderEx.cs]: These is no tests as far as I can see in the reference implementation

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs (hunk 1 above)
Comment (mregen) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs:1735]: Interesting, I will take a look.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Models/EnumDefinition2.cs [hunk 4]
```diff
@@ -25,11 +25,6 @@ public class EnumDefinition2 : EnumDefinition {
         /// </summary>
         public LocalizedText Description { get; internal set; }
 
-        /// <summary>
-        /// Whether the enum is an option set
-        /// </summary>
-        public bool IsOptionSet { get; set; }
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Models/EnumDefinition2.cs:31]: >IsOptionSet [](http://example.com/codeflow?start=20&length=11)

ist this already defined in the base class?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs [hunk 5]
```diff
@@ -1731,7 +1732,7 @@ private class DummyDialog : IApplicationMessageDlg {
...
         }
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs:1736]: >} [](http://example.com/codeflow?start=8&length=1)

implement /// <summary>
        /// Creates an instance of the service host.
        /// </summary>
        public virtual ServiceHost CreateServiceHost(ServerBase server, params Uri[] addresses)
        {
            return new ServiceHost()
        }

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs (hunk 1 above)
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs:1735]: call serverBase.CreateServiceHost etc..

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs [hunk 6]
```diff
@@ -1731,7 +1737,7 @@ private class DummyDialog : IApplicationMessageDlg {
                 DiscoveryUrls = null
             };
             endpoints = _endpoints;
-            return new List<Task> { Task.CompletedTask };
+            return new List<ServiceHost> { CreateServiceHost(this, null) };
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs:1740]: >null [](http://example.com/codeflow?start=67&length=4)

nth, Uri from endpoints

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj [hunk 7]
```diff
@@ -42,20 +42,20 @@
...
     <When Condition="'$(Configuration)'=='Release'">
       <ItemGroup>
-        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Core" Version="1.4.365-gfc341ee8c5" />
-        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Configuration" Version="1.4.365-gfc341ee8c5" />
-        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Server" Version="1.4.365-gfc341ee8c5" />
-        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Client" Version="1.4.365-gfc341ee8c5" />
-        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Client.ComplexTypes" Version="1.4.365-gfc341ee8c5" />
+        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Core" Version="1.4.366.31-preview" />
+        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Configuration" Version="1.4.366.31-preview" />
+        <PackageReference Include="OPCFoundation.NetStandard.Opc.Ua.Server" Version="1.4.366.31-preview" />
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj:47]: >rsion="1.4.366.31-pre [](http://example.com/codeflow?start=78&length=21)

366 release will be ready end of day, but one more small change in the IEncoder ...

**Review by mregen — APPROVED** [review]
:shipit:

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs (hunk 6 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Gateway/src/Server/GatewayServer.cs]: Sorry, sloppy. Fixed.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj (hunk 7 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj:47]: ok, let's wait for that

**Comment by mregen** [comment]
366.38 is out...

**Comment by peterbryntesson** [comment]
> 366.38 is out...

Back from OneWeek now - restarting work on this by upgrading package.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Models/EnumDefinition2.cs (hunk 4 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Stack/Models/EnumDefinition2.cs:31]: Yes.
image

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj (hunk 7 above)
Comment (peterbryntesson) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Protocol/src/Microsoft.Azure.IIoT.OpcUa.Protocol.csproj]: Upgraded packages to 1.4.366.38

**Review by simonjaeger — APPROVED** [review]

**Pull request merged**